cosmwasm/contracts/ibc-translator/src/reply.rs

```diff
@@ -0,0 +1,197 @@
+
+    // deserialize payload into the type we expect
+    let payload: GatewayIbcTokenBridgePayload = serde_json_wasm::from_slice(&transfer_info.payload)
+        .context("failed to deserialize transfer payload")?;
+
+    match payload {
+        GatewayIbcTokenBridgePayload::Simple {
+            chain,
+            recipient,
+            fee: _,
+            nonce: _,
+        } => {
+            let recipient_decoded = String::from_utf8(recipient.to_vec())
+                .context(format!("failed to convert {recipient} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                recipient_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                None,
+            )
+        }
+        GatewayIbcTokenBridgePayload::ContractControlled {
+            chain,
+            contract,
+            payload,
+            nonce: _,
+        } => {
+            let contract_decoded = String::from_utf8(contract.to_vec())
+                .context(format!("failed to convert {contract} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                contract_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                Some(payload),
+            )
+        }
+    }
+}
+pub fn convert_cw20_to_bank_and_send(
+    deps: DepsMut,
+    env: Env,
+    recipient: String,
+    amount: u128,
+    contract_addr: String,
+    chain_id: u16,
+    payload: Option<Binary>,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    // check the recipient and contract addresses are valid
+    // recipient will have a different bech32 prefix and fail
+    //deps.api
+    //    .addr_validate(&recipient)
+    //    .context(format!("invalid recipient address {}", recipient))?;
+
+    deps.api
+        .addr_validate(&contract_addr)
+        .context(format!("invalid contract address {contract_addr}"))?;
+
+    // convert contract address into base64
+    let subdenom = contract_addr_to_base58(deps.as_ref(), contract_addr.clone())?;
+    // format the token factory denom
+    let tokenfactory_denom = "factory/".to_string()
+        + env.contract.address.to_string().as_ref()
+        + "/"
+        + subdenom.as_ref();
+
+    let mut response: Response<TokenFactoryMsg> = Response::new();
+
+    // check contract storage see if we've created a denom for this cw20 token yet
+    // if we haven't created the denom, then create the denom
+    // info.sender contains the cw20 contract address
+    if !CW_DENOMS.has(deps.storage, contract_addr.clone()) {
+        // call into token factory to create the denom
+        let create_denom = SubMsg::reply_on_success(
+            TokenMsg::CreateDenom {
+                subdenom,
+                metadata: None,
+            },
+            CREATE_DENOM_REPLY_ID,
```

Comment:
why do we need a reply handler for the `CreateDenom` call?

Reply:
We don't, I removed it